zkevm-circuits/src/evm_circuit/op_execution/comparator.rs

```diff
@@ -1,379 +1,196 @@
+            &mut cb,
+            state_curr.opcode.expr(),
+            OpcodeId::GT.expr(),
+        );
 
-        let out_of_gas = {
-            let (selector, gas_available) = &self.out_of_gas;
-            let gas_overdemand = state_curr.gas_counter.expr()
-                + GasCost::FASTEST.expr()
-                - gas_available.expr();
-            Constraint {
-                name: "LtGadget out of gas",
-                selector: selector.expr(),
-                polys: vec![
-                    (gas_overdemand.clone() - 1.expr())
-                        * (gas_overdemand.clone() - 2.expr())
-                        * (gas_overdemand - 3.expr()),
-                ],
-                lookups: vec![],
-            }
-        };
+        // `a[0..16] <= b[0..16]`
+        let (lt_lo, eq_lo) = self.comparison_lo.constraints(
+            &mut cb,
+            from_bytes::expr(self.a.cells[0..16].to_vec()),
+            from_bytes::expr(self.b.cells[0..16].to_vec()),
+        );
 
-        array::IntoIter::new([success, stack_underflow, out_of_gas])
-            .map(move |mut constraint| {
-                constraint.polys =
-                    [common_polys.clone(), constraint.polys].concat();
-                constraint
-            })
-            .collect()
-    }
+        // `a[16..32] <= b[16..32]`
+        let (lt_hi, eq_hi) = self.comparison_hi.constraints(
+            &mut cb,
+            from_bytes::expr(self.a.cells[16..32].to_vec()),
+            from_bytes::expr(self.b.cells[16..32].to_vec()),
+        );
 
-    fn assign(
-        &self,
-        region: &mut Region<'_, F>,
-        offset: usize,
-        core_state: &mut CoreStateInstance,
-        execution_step: &ExecutionStep,
-    ) -> Result<(), Error> {
-        match execution_step.case {
-            Case::Success => {
-                self.assign_success(region, offset, core_state, execution_step)
-            }
-            Case::StackUnderflow => {
-                unimplemented!()
-            }
-            Case::OutOfGas => {
-                unimplemented!()
-            }
-            _ => unreachable!(),
+        // `a < b` when:
+        // - `a[16..32] < b[16..32]` OR
+        // - `a[16..32] == b[16..32]` AND `a[0..16] < b[0..16]`
+        let lt = select::expr(lt_hi, 1.expr(), eq_hi.clone() * lt_lo);
+        // `a == b` when both parts are equal
+        let eq = eq_hi * eq_lo;
+
+        // The result is:
+        // - `lt` when LT or GT
+        // - `eq` when EQ
+        let result = select::expr(is_eq_op, eq, lt);
+
+        // Pop a and b from the stack, push the result on the stack.
+        // When swap is enabled we swap stack places between a and b.
+        // We can push result here directly because
+        // it only uses the LSB of a word.
+        cb.stack_pop(select::expr(swap.clone(), self.b.expr(), self.a.expr()));
+        cb.stack_pop(select::expr(swap, self.a.expr(), self.b.expr()));
+        cb.stack_push(result);
```

Comment:
So this can be correct only when the value is less than 256, right?

Reply:
Yep, see Han's comment here: https://github.com/appliedzkp/zkevm-circuits/pull/120#discussion_r738913187